### DISCUSSION TITLE
What is blocking production use?

### DISCUSSION THREAD

**Discussion by nibanks** [microsoft/windows-drivers-rs#11] [Q&A] [answered]
The readme currently state:

> Note: This project is still in early stages of development and is not yet recommended for commercial use.

Our team (Windows Networking) has several open-source drivers ([microsoft/msquic](https://github.com/microsoft/msquic), [microsoft/xdp-for-windows](https://github.com/microsoft/xdp-for-windows), [microsoft/ebpf-for-windows](https://github.com/microsoft/ebpf-for-windows)) currently in C that might be able to benefit from this project, but we'd obviously need "commercial use". So, I was wondering what is currently blocking commercial use? Is there a place tracking what is left that we can follow along?

**Comment by gplusplus314** [comment, marked as answer]
Hey there. I landed here looking to ask a similar question. While perusing the repo, I came across #22, which changes the verbiage from “commercial” to “production,” with all the usual disclaimers about using experimental software.

**Comment by JohnAZoidberg** [comment]
@wmmc88 what's the current status?
Is it okay to build production drivers with this now?
Would Microsoft sign drivers built on top of this?

**Reply by wmmc88** [reply]
The statement in the readme still stands. This repo is open source and you may choose to do with it what you want. In order to ship a driver for Windows, you still need to follow all the standard procedures required for HLK and WHQL, and need to go through the normal Windows Hardware Dev Center processes. Off the top of my head, you may run into issues with [Static Tools Logo Test](https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/static-tools-and-codeql) since it looks like [CodeQL does not support Rust yet](https://codeql.github.com/docs/codeql-overview/supported-languages-and-frameworks/). If you have more questions about that, I'd recommend you reach out to those respective teams through their respective channels.

**Reply by GF-Huang** [reply]
> Off the top of my head, you may run into issues with [Static Tools Logo Test](https://learn.microsoft.com/en-us/windows-hardware/drivers/devtest/static-tools-and-codeql) since it looks like [CodeQL does not support Rust yet](https://codeql.github.com/docs/codeql-overview/supported-languages-and-frameworks/). If you have more questions about that, I'd recommend you reach out to those respective teams through their respective channels.

So it's still a toy rather than a production tool for now?

**Reply by era** [reply]
I appreciate that there are other reasons for not considering it production-ready, but I would like to highlight that [CodeQL supports Rust now](https://codeql.github.com/docs/codeql-overview/supported-languages-and-frameworks/) ("Features from nightly toolchains are not supported.").

**Reply by jdoubleu** [reply]
This paper, as well as other reports (e.g. https://zaddach.github.io/2025-11-10-windows-drivers-in-rust/ (see https://github.com/microsoft/windows-drivers-rs/discussions/564)) might be helpful when making a decision.
> Cot, C., Portase, R., Coleşa, A. (2026). Rust in the Kernel: A Practical Evaluation on Windows. In: D'Arco, P., Zamfiroiu, A. (eds) Innovative Security Solutions for Information Technology and Communications. SecITC 2025. Lecture Notes in Computer Science, vol 16443. Springer, Cham. https://doi.org/10.1007/978-3-032-17443-7_19
